crates/dbsp/src/algebra/decimal.rs

```diff
@@ -0,0 +1,607 @@
+use num::{FromPrimitive, Num, One, Signed, ToPrimitive, Zero};
+use rust_decimal::{Decimal, MathematicalOps};
+use size_of::SizeOf;
+use std::hash::Hash;
+use std::{
+    fmt::Display,
+    ops::{Add, AddAssign, Div, Mul, Neg, Rem, Sub},
+    str::FromStr,
+};
+
+mod inner_types {
+    use rust_decimal::Decimal;
+    use size_of::SizeOf;
+    use std::fmt::Display;
+
+    const MIN_DECIMAL_PRECISION: u32 = 1;
+    const MAX_DECIMAL_PRECISION: u32 = 38;
+
+    const MIN_DECIMAL_SCALE: u32 = 0;
+    const MAX_DECIMAL_SCALE: u32 = 28;
+
+    #[derive(
+        Debug,
+        Clone,
+        Copy,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Hash,
+        SizeOf,
+        rkyv::Archive,
+        rkyv::Serialize,
+        rkyv::Deserialize,
+    )]
+    #[cfg_attr(feature = "with-serde", derive(serde::Serialize, serde::Deserialize))]
+    pub struct Precision(u32);
+
+    impl Precision {
+        pub fn new(val: Decimal) -> Result<Self, String> {
+            let mut precision = val.mantissa().checked_ilog10().unwrap_or(0);
+
+            if precision == 0 {
+                precision = val.scale() + 1;
+            }
+
+            Self::try_from(precision)
+        }
+
+        pub fn inner(&self) -> u32 {
+            self.0
+        }
+
+        pub fn max() -> Self {
+            Self(MAX_DECIMAL_PRECISION)
+        }
+    }
+
+    impl TryFrom<u32> for Precision {
+        type Error = String;
+
+        fn try_from(value: u32) -> Result<Self, Self::Error> {
+            match value {
+                x if (MIN_DECIMAL_PRECISION..=MAX_DECIMAL_PRECISION).contains(&x) => Ok(Precision(x)),
+                _ => Err(format!(
+                    "Numeric precision {value} must be between {MIN_DECIMAL_PRECISION} and {MAX_DECIMAL_PRECISION}"
+                )),
+            }
+        }
+    }
+
+    impl Display for Precision {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "{}", self.0)
+        }
+    }
+
+    impl Default for Precision {
+        fn default() -> Self {
+            Precision(MAX_DECIMAL_PRECISION)
+        }
+    }
+
+    // In Postgres, the scale can be between [-1000, 1000]
+    #[derive(
+        Debug,
+        Clone,
+        Copy,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Hash,
+        SizeOf,
+        rkyv::Archive,
+        rkyv::Serialize,
+        rkyv::Deserialize,
+    )]
+    #[cfg_attr(feature = "with-serde", derive(serde::Serialize, serde::Deserialize))]
+    pub struct Scale(u32);
```

Comment:
If you plan to support negative scales in the future perhaps you should make this an i32.